### PR TITLE
[WIP] Support YOLOX

### PR DESCRIPTION
**Pull request by hhaAndroid**
We plan to support YOLOX in the near future. The plan is divided into 3 steps：

1. Inference accuracy alignment
2. Training accuracy alignment
3. Refactor

## step 1
Inference accuracy alignment.

| model   | official(val2017) | mmdet(val2017) |
| ------- | ----------------- | -------------- |
| yolox_s | 39.2              | 39.2           |
| yolox_tiny | 31.9              | 31.9           |

@hhaAndroid 

## step2
Training accuracy alignment.

-  [x]  dataset @HAOCHENYE
-  [x]  lr_scheduler @zhiqwang
-  [x]  optimizer @zytx121
-  [x]  loss @ 
-  [x]   ema @hhaAndroid 
-  [x]  other @HsLOL

Done!

## step3
Refactor. Please refer to https://github.com/open-mmlab/mmdetection/pull/5692, https://github.com/open-mmlab/mmdetection/pull/5669

### CODE REVIEW

File: mmdet/models/utils/block.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        self.act = get_activation(act, inplace=True)
+
+    def forward(self, x):
+        return self.act(self.bn(self.conv(x)))
+
+    def fuseforward(self, x):
+        return self.act(self.conv(x))
+
+
+class DWConv(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/models/utils/block.py:51]: We have Depthwise ConvModule, see https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/depthwise_separable_conv_module.py

File: mmdet/models/utils/block.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    elif name == "relu":
+        module = nn.ReLU(inplace=inplace)
+    elif name == "lrelu":
+        module = nn.LeakyReLU(0.1, inplace=inplace)
+    else:
+        raise AttributeError("Unsupported act type: {}".format(name))
+    return module
+
+
+class BaseConv(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/models/utils/block.py:25]: Can use convmodule

File: mmdet/models/utils/block.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        self.conv2 = Conv(hidden_channels, out_channels, 3, stride=1, act=act)
+        self.use_add = shortcut and in_channels == out_channels
+
+    def forward(self, x):
+        y = self.conv2(self.conv1(x))
+        if self.use_add:
+            y = y + x
+        return y
+
+class Focus(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/models/utils/block.py:140]: Spatial2Channel might be better

File: mmdet/apis/ema.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+from mmcv.runner.hooks import HOOKS, Hook
```
Comment (ZwwWayne) [inline mmdet/apis/ema.py:1]: It is better to put this file into core/hooks/

File: mmdet/datasets/mosaic_dataset.py [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+        im_ann = self.coco.loadImgs(id_)[0]
+        width = im_ann["width"]
+        height = im_ann["height"]
+
+        # load labels
+        valid_objs = []
+        for obj in annotations:
+            x1 = np.max((0, obj["bbox"][0]))
+            y1 = np.max((0, obj["bbox"][1]))
+            x2 = np.min((width - 1, x1 + np.max((0, obj["bbox"][2] - 1))))
```
Comment (ZwwWayne) [inline mmdet/datasets/mosaic_dataset.py:284]: This is inconsistent with the coordinate system.
